### PR TITLE
qa: Adding rgw multisite support testing

### PR DESCRIPTION
**Pull request by rkachach**
## Contribution Guidelines
- To sign and title your commits, please refer to [Submitting Patches to Ceph](https://github.com/ceph/ceph/blob/main/SubmittingPatches.rst).

- If you are submitting a fix for a stable branch (e.g. "pacific"), please refer to [Submitting Patches to Ceph - Backports](https://github.com/ceph/ceph/blob/master/SubmittingPatches-backports.rst) for the proper workflow.

## Checklist
- Tracker (select at least one)
  - [x] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [ ] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [x] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [ ] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [x] No doc update is appropriate
- Tests (select at least one)
  - [ ] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [x] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`
- `jenkins test windows`
</details>

### CODE REVIEW

File: qa/suites/orch/cephadm/workunits/task/test_rgw_multisite.yaml [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+      - ceph mgr module enable rgw
+- rgw_module.apply:
+    specs:
+      - rgw_realm: myrealm1
+        rgw_zonegroup: myzonegroup1
+        rgw_zone: myzone1
+        placement:
+          hosts:
+           - ceph-node-0
+           - ceph-node-1
```
Comment (adk3798) [inline qa/suites/orch/cephadm/workunits/task/test_rgw_multisite.yaml:27]:
```
2023-03-14T12:27:30.260 INFO:journalctl@ceph.mon.a.smithi064.stdout:Mar 14 12:27:29 smithi064 ceph-mon[94293]: osdmap e32: 3 total, 3 up, 3 in
2023-03-14T12:27:30.266 INFO:teuthology.orchestra.run.smithi064.stderr:Error EINVAL: Traceback (most recent call last):
2023-03-14T12:27:30.266 INFO:teuthology.orchestra.run.smithi064.stderr:  File "/usr/share/ceph/mgr/mgr_module.py", line 1763, in _handle_command
2023-03-14T12:27:30.266 INFO:teuthology.orchestra.run.smithi064.stderr:    return CLICommand.COMMANDS[cmd['prefix']].call(self, cmd, inbuf)
2023-03-14T12:27:30.266 INFO:teuthology.orchestra.run.smithi064.stderr:  File "/usr/share/ceph/mgr/mgr_module.py", line 462, in call
2023-03-14T12:27:30.266 INFO:teuthology.orchestra.run.smithi064.stderr:    return self.func(mgr, **kwargs)
2023-03-14T12:27:30.266 INFO:teuthology.orchestra.run.smithi064.stderr:  File "/usr/share/ceph/mgr/rgw/module.py", line 96, in wrapper
2023-03-14T12:27:30.267 INFO:teuthology.orchestra.run.smithi064.stderr:    return func(self, *args, **kwargs)
2023-03-14T12:27:30.267 INFO:teuthology.orchestra.run.smithi064.stderr:  File "/usr/share/ceph/mgr/rgw/module.py", line 189, in _cmd_rgw_realm_bootstrap
2023-03-14T12:27:30.267 INFO:teuthology.orchestra.run.smithi064.stderr:    RGWAM(self.env).realm_bootstrap(spec, start_radosgw)
2023-03-14T12:27:30.267 INFO:teuthology.orchestra.run.smithi064.stderr:  File "/lib/python3.6/site-packages/ceph/rgw/rgwam_core.py", line 559, in realm_bootstrap
2023-03-14T12:27:30.267 INFO:teuthology.orchestra.run.smithi064.stderr:    self.env.mgr.apply_rgw(rgw_spec)
2023-03-14T12:27:30.267 INFO:teuthology.orchestra.run.smithi064.stderr:  File "/usr/share/ceph/mgr/rgw/module.py", line 76, in apply_rgw
2023-03-14T12:27:30.267 INFO:teuthology.orchestra.run.smithi064.stderr:    orchestrator.raise_if_exception(completion)
2023-03-14T12:27:30.267 INFO:teuthology.orchestra.run.smithi064.stderr:  File "/usr/share/ceph/mgr/orchestrator/_interface.py", line 228, in raise_if_exception
2023-03-14T12:27:30.268 INFO:teuthology.orchestra.run.smithi064.stderr:    raise e
2023-03-14T12:27:30.268 INFO:teuthology.orchestra.run.smithi064.stderr:orchestrator._interface.OrchestratorValidationError: Cannot place <RGWSpec for service_name=rgw.myrealm1.myzone1> on ceph-node-0, ceph-node-1: Unknown hosts
2023-03-14T12:27:30.268 INFO:teuthology.orchestra.run.smithi064.stderr:
2023-03-14T12:27:30.759 INFO:journalctl@ceph.mgr.a.smithi064.stdout:Mar 14 12:27:30 smithi064 conmon[94718]: 2023-03-14T12:27:30.258+0000 7f6316d1d700 -1 log_channel(cephadm) log [ERR] : Cannot place <RGWSpec for service_name=rgw.myrealm1.myzone1> on ceph-node-0, ceph-node-1: Unknown hosts
```
@rkachach I don't think explicit hosts like this are going to work in the teuthology test

Comment (adk3798) [reply qa/suites/orch/cephadm/workunits/task/test_rgw_multisite.yaml:27]: or if we really need them, you'd have to first grab the hostnames from `ceph orch host ls` and inject them into the spec, which I think is possible but probably annoying to do

Comment (adk3798) [reply qa/suites/orch/cephadm/workunits/task/test_rgw_multisite.yaml:27]: https://pulpito.ceph.com/adking-2023-03-14_05:03:12-orch:cephadm-wip-adk-testing-2023-03-13-1654-distro-default-smithi/7206254/

Comment (rkachach) [reply qa/suites/orch/cephadm/workunits/task/test_rgw_multisite.yaml]: @adk3798 I removed the whole placement section as in this case it doesn't really matter where the daemons are placed.

File: qa/suites/orch/cephadm/workunits/task/test_rgw_multisite.yaml [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+          rgw_frontend_port: 5500
+- cephadm.shell:
+    host.a:
+      - |
+        set -e
+        set -x
+        TOKENS=$(ceph rgw realm tokens)
+        echo $TOKENS | jq --exit-status '.[0].realm == "myrealm1"'
+        echo $TOKENS | jq --exit-status '.[0].token'
+        TOKEN_JSON=$(ceph rgw realm tokens | jq -r '.[0].token' | base64 --decode)
```
Comment (adk3798) [inline qa/suites/orch/cephadm/workunits/task/test_rgw_multisite.yaml:34]: Test doesn't seem to quite work yet. I think it's failing here
```
2023-03-22T15:14:28.729 INFO:teuthology.orchestra.run.smithi019.stderr:+ TOKENS='[
2023-03-22T15:14:28.729 INFO:teuthology.orchestra.run.smithi019.stderr:    {
2023-03-22T15:14:28.729 INFO:teuthology.orchestra.run.smithi019.stderr:        "realm": "myrealm1",
2023-03-22T15:14:28.729 INFO:teuthology.orchestra.run.smithi019.stderr:        "token": "master zone has no endpoint"
2023-03-22T15:14:28.730 INFO:teuthology.orchestra.run.smithi019.stderr:    }
2023-03-22T15:14:28.730 INFO:teuthology.orchestra.run.smithi019.stderr:]'
2023-03-22T15:14:28.730 INFO:teuthology.orchestra.run.smithi019.stderr:+ echo '[' '{' '"realm":' '"myrealm1",' '"token":' '"master' zone has no 'endpoint"' '}' ']'
2023-03-22T15:14:28.730 INFO:teuthology.orchestra.run.smithi019.stderr:+ jq --exit-status '.[0].realm == "myrealm1"'
2023-03-22T15:14:28.760 INFO:teuthology.orchestra.run.smithi019.stdout:true
2023-03-22T15:14:28.761 INFO:teuthology.orchestra.run.smithi019.stderr:+ echo '[' '{' '"realm":' '"myrealm1",' '"token":' '"master' zone has no 'endpoint"' '}' ']'
2023-03-22T15:14:28.761 INFO:teuthology.orchestra.run.smithi019.stderr:+ jq --exit-status '.[0].token'
2023-03-22T15:14:28.791 INFO:teuthology.orchestra.run.smithi019.stdout:"master zone has no endpoint"
2023-03-22T15:14:28.792 INFO:teuthology.orchestra.run.smithi019.stderr:++ ceph rgw realm tokens
2023-03-22T15:14:28.792 INFO:teuthology.orchestra.run.smithi019.stderr:++ jq -r '.[0].token'
2023-03-22T15:14:28.792 INFO:teuthology.orchestra.run.smithi019.stderr:++ base64 --decode
2023-03-22T15:14:29.424 INFO:journalctl@ceph.mon.c.smithi202.stdout:Mar 22 15:14:29 smithi202 ceph-mon[96864]: Deploying daemon rgw.myrealm1.myzone1.smithi202.bswwyw on smithi202
2023-03-22T15:14:29.425 INFO:journalctl@ceph.mon.c.smithi202.stdout:Mar 22 15:14:29 smithi202 ceph-mon[96864]: from='client.14730 -' entity='client.admin' cmd=[{"prefix": "rgw realm tokens", "target": ["mon-mgr", ""]}]: dispatch
2023-03-22T15:14:29.429 INFO:teuthology.orchestra.run.smithi019.stderr:base64: invalid input
2023-03-22T15:14:29.430 INFO:teuthology.orchestra.run.smithi019.stderr:+ TOKEN_JSON=$'\231\253-z'
2023-03-22T15:14:29.445 INFO:journalctl@ceph.mon.a.smithi019.stdout:Mar 22 15:14:29 smithi019 ceph-mon[94540]: Deploying daemon rgw.myrealm1.myzone1.smithi202.bswwyw on smithi202
2023-03-22T15:14:29.446 INFO:journalctl@ceph.mon.a.smithi019.stdout:Mar 22 15:14:29 smithi019 ceph-mon[94540]: from='client.14730 -' entity='client.admin' cmd=[{"prefix": "rgw realm tokens", "target": ["mon-mgr", ""]}]: dispatch
2023-03-22T15:14:29.621 INFO:journalctl@ceph.mon.b.smithi090.stdout:Mar 22 15:14:29 smithi090 ceph-mon[96538]: Deploying daemon rgw.myrealm1.myzone1.smithi202.bswwyw on smithi202
2023-03-22T15:14:29.621 INFO:journalctl@ceph.mon.b.smithi090.stdout:Mar 22 15:14:29 smithi090 ceph-mon[96538]: from='client.14730 -' entity='client.admin' cmd=[{"prefix": "rgw realm tokens", "target": ["mon-mgr", ""]}]: dispatch
2023-03-22T15:14:30.135 DEBUG:teuthology.orchestra.run:got remote process result: 1
```
https://pulpito.ceph.com/adking-2023-03-15_01:34:51-orch:cephadm-wip-adk-testing-2023-03-14-1905-distro-default-smithi/7208451

Comment (rkachach) [reply qa/suites/orch/cephadm/workunits/task/test_rgw_multisite.yaml:34]: @adk3798 it seems that when the checks are ran, the rgw daemons are not created yet. Not sure if we should add some active-wait or in fact the rgw are not being created for some other reason.

**Comment by adk3798** [comment]
at least one instance of this test passed in the recent run https://pulpito.ceph.com/adking-2023-03-30_12:19:06-orch:cephadm-wip-adk-testing-2023-03-28-1716-distro-default-smithi/7226195

**Comment by rkachach** [comment]
> at least one instance of this test passed in the recent run https://pulpito.ceph.com/adking-2023-03-30_12:19:06-orch:cephadm-wip-adk-testing-2023-03-28-1716-distro-default-smithi/7226195

Thanks @adk3798. From the logs it seems that the checks are being processed correctly which is a good thing:

```
2023-03-30T12:51:21.230 INFO:teuthology.orchestra.run.smithi081.stderr:++ ceph rgw realm tokens
2023-03-30T12:51:21.230 INFO:teuthology.orchestra.run.smithi081.stderr:++ jq -r '.[0].token'
2023-03-30T12:51:21.230 INFO:teuthology.orchestra.run.smithi081.stderr:++ base64 --decode
2023-03-30T12:51:21.814 INFO:teuthology.orchestra.run.smithi081.stderr:+ TOKEN_JSON='{
2023-03-30T12:51:21.814 INFO:teuthology.orchestra.run.smithi081.stderr:    "realm_name": "myrealm1",
2023-03-30T12:51:21.814 INFO:teuthology.orchestra.run.smithi081.stderr:    "realm_id": "99548192-2933-4af5-9b0e-a109bbbef76c",
2023-03-30T12:51:21.815 INFO:teuthology.orchestra.run.smithi081.stderr:    "endpoint": "http://smithi132:5500",
2023-03-30T12:51:21.815 INFO:teuthology.orchestra.run.smithi081.stderr:    "access_key": "K71AUJQLBX7386ROUAK1",
2023-03-30T12:51:21.815 INFO:teuthology.orchestra.run.smithi081.stderr:    "secret": "ECJFqqDpWkboc4yAsM4LcTl1koh3pLXeX1zfOzYt"
2023-03-30T12:51:21.815 INFO:teuthology.orchestra.run.smithi081.stderr:}'
2023-03-30T12:51:21.816 INFO:teuthology.orchestra.run.smithi081.stderr:+ echo '{' '"realm_name":' '"myrealm1",' '"realm_id":' '"99548192-2933-4af5-9b0e-a109bbbef76c",' '"endpoint":' '"http://smithi132:5500",' '"access_key":' '"K71AUJQLBX7386ROUAK1",' '"secret":' '"ECJFqqDpWkboc4yAsM4LcTl1koh3pLXeX1zfOzYt"' '}'
2023-03-30T12:51:21.816 INFO:teuthology.orchestra.run.smithi081.stderr:+ jq --exit-status '.realm_name == "myrealm1"'
2023-03-30T12:51:21.845 INFO:teuthology.orchestra.run.smithi081.stdout:true
2023-03-30T12:51:21.846 INFO:teuthology.orchestra.run.smithi081.stderr:+ echo '{' '"realm_name":' '"myrealm1",' '"realm_id":' '"99548192-2933-4af5-9b0e-a109bbbef76c",' '"endpoint":' '"http://smithi132:5500",' '"access_key":' '"K71AUJQLBX7386ROUAK1",' '"secret":' '"ECJFqqDpWkboc4yAsM4LcTl1koh3pLXeX1zfOzYt"' '}'
2023-03-30T12:51:21.847 INFO:teuthology.orchestra.run.smithi081.stderr:+ jq --exit-status '.endpoint | test("http://.+:\\d+")'
2023-03-30T12:51:21.876 INFO:teuthology.orchestra.run.smithi081.stdout:true
2023-03-30T12:51:21.877 INFO:teuthology.orchestra.run.smithi081.stderr:+ echo '{' '"realm_name":' '"myrealm1",' '"realm_id":' '"99548192-2933-4af5-9b0e-a109bbbef76c",' '"endpoint":' '"http://smithi132:5500",' '"access_key":' '"K71AUJQLBX7386ROUAK1",' '"secret":' '"ECJFqqDpWkboc4yAsM4LcTl1koh3pLXeX1zfOzYt"' '}'
2023-03-30T12:51:21.877 INFO:teuthology.orchestra.run.smithi081.stderr:+ jq --exit-status '.realm_id | test("^[0-9a-f]{8}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{12}$")'
2023-03-30T12:51:21.907 INFO:teuthology.orchestra.run.smithi081.stdout:true
2023-03-30T12:51:21.908 INFO:teuthology.orchestra.run.smithi081.stderr:+ echo '{' '"realm_name":' '"myrealm1",' '"realm_id":' '"99548192-2933-4af5-9b0e-a109bbbef76c",' '"endpoint":' '"http://smithi132:5500",' '"access_key":' '"K71AUJQLBX7386ROUAK1",' '"secret":' '"ECJFqqDpWkboc4yAsM4LcTl1koh3pLXeX1zfOzYt"' '}'
2023-03-30T12:51:21.909 INFO:teuthology.orchestra.run.smithi081.stderr:+ jq --exit-status .access_key
2023-03-30T12:51:21.938 INFO:teuthology.orchestra.run.smithi081.stdout:"K71AUJQLBX7386ROUAK1"
2023-03-30T12:51:21.939 INFO:teuthology.orchestra.run.smithi081.stderr:+ echo '{' '"realm_name":' '"myrealm1",' '"realm_id":' '"99548192-2933-4af5-9b0e-a109bbbef76c",' '"endpoint":' '"http://smithi132:5500",' '"access_key":' '"K71AUJQLBX7386ROUAK1",' '"secret":' '"ECJFqqDpWkboc4yAsM4LcTl1koh3pLXeX1zfOzYt"' '}'
2023-03-30T12:51:21.939 INFO:teuthology.orchestra.run.smithi081.stderr:+ jq --exit-status .secret
```

**Review by adk3798 — APPROVED** [review]

**Comment by adk3798** [comment]
https://pulpito.ceph.com/adking-2023-04-04_02:06:27-orch:cephadm-wip-adk-testing-2023-04-03-1541-distro-default-smithi/

Reruns of failed/dead jobs:  https://pulpito.ceph.com/adking-2023-04-04_05:19:12-orch:cephadm-wip-adk-testing-2023-04-03-1541-distro-default-smithi/

After reruns, 6 failures. All 6 failures were on tests being added/adjusted by PRs in the run (2 keepalive-only nfs tests, 2 mon set crush location tests, 2 staggered upgrade tests) and those failures don't seem like they should block merging for any PRs in the run outside of those adjusting the tests.

**Pull request merged**